src/python/gudhi/point_cloud/timedelay.py

```diff
@@ -0,0 +1,56 @@
+import numpy as np
+
+class TimeDelayEmbedding:
+    """Point cloud transformation class.
+
+    Embeds time-series data in the R^d according to Takens' Embedding Theorem
+    and obtains the coordinates of each point.
+
+    Parameters
+    ----------
+    dim : int, optional (default=3)
+        `d` of R^d to be embedded.
+
+    delay : int, optional (default=1)
+        Time-Delay embedding.
+
+    skip : int, optional (default=1)
+        How often to skip embedded points.
+
+    """
+    def __init__(self, dim=3, delay=1, skip=1):
+        self._dim = dim
+        self._delay = delay
+        self._skip = skip
+
+    def __call__(self, *args, **kwargs):
+        return self.transform(*args, **kwargs)
+
+    def _transform(self, ts):
+        """Guts of transform method."""
+        return ts[
+            np.add.outer(
+                np.arange(0, len(ts)-self._delay*(self._dim-1), self._skip),
+                np.arange(0, self._dim*self._delay, self._delay))
+        ]
+
+    def transform(self, ts):
+        """Transform method.
+
+        Parameters
+        ----------
+        ts : list[float] or list[list[float]]
+            A single or multiple time-series data.
```

Comment:
Asking for opinions: How about splitting the functionality for 1 vs multiple series into several functions? Say transform would only work on multiple series, because that's what sklearn expects, while `__call__` would only work on a single series? Or some other similar scheme.

Reply:
An advantage of separating single vs multiple is that it would let us generalize the function, so it can take as input say a (n,2) array representing a time series with a 2D vector at each time, and it can return a (n-2,6) array where we take 3 consecutive 2D vectors and concatenate them into a 6D vector. This is already what `_transform` does, up to a final call to `reshape`.

Reply:
Thank you for valuable comment.  
`__call__`, `_transform` and `transform` have been modified as follows.  
Is code good like this?  
  
- Before  
```
def __call__(self, *args, **kwargs):
    return self.transform(*args, **kwargs)
```
- After  
```
def __call__(self, *args, **kwargs):
    return self._transform(*args, **kwargs)
```  

- Before
```
def _transform(self, ts):
    """Guts of transform method.

    Parameters
    ----------
    ts : ndarray, shape=[n_features]
        A single time-series data.

    Returns
    -------
    attractor : ndarray, shape=[n_points, d]
        `n_points` is the number of points in the R^d.

    """
    return ts[
        np.add.outer(
            np.arange(0, len(ts)-self._delay*(self._dim-1), self._skip),
            np.arange(0, self._dim*self._delay, self._delay))
    ]
```
- After  
```
def _transform(self, ts):
    """Guts of transform method.

    Parameters
    ----------
    ts : ndarray, shape=[n_features]
        A single time-series data.

    Returns
    -------
    attractor : ndarray, shape=[n_points, d]
        `n_points` is the number of points in the R^d.

    """
    ts = np.array(ts)
    return ts[
        np.add.outer(
            np.arange(0, len(ts)-self._delay*(self._dim-1), self._skip),
            np.arange(0, self._dim*self._delay, self._delay))
    ]
```
- Before
```
def transform(self, ts):
    """Transform method."""
    ndts = np.array(ts)
    if ndts.ndim == 1:
        # for single.
        return self._transform(ndts)
    else:
        # for multiple.
        return np.apply_along_axis(self._transform, 1, ndts)
```  
- After
```   
# for multiple
def transform(self, ts):
    """Transform method."""
    ndts = np.array(ts)
    return np.apply_along_axis(self._transform, 1, ndts)
```

Reply:
Yes, that looks good to me, if you agree with that interface.

Reply:
The following has been reflected in the code.  
Please comment.
- splitting the functionality for 1 vs multiple series  
- ndts = np.array(ts) means  
- Fix delay and skip example
- add fit (self, X, y = None)  
  
```
class TimeDelayEmbedding:
    """Point cloud transformation class.
    Embeds time-series data in the R^d according to Takens' Embedding Theorem
    and obtains the coordinates of each point.
    Parameters
    ----------
    dim : int, optional (default=3)
        `d` of R^d to be embedded.
    delay : int, optional (default=1)
        Time-Delay embedding.
    skip : int, optional (default=1)
        How often to skip embedded points.
        Given delay=3 and skip=2, an point cloud which is obtained by embedding 
        a single time-series data into R^3 is as follows.
        
    .. code-block:: none
    
    time-series = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    point clouds = [[1, 4, 7], 
                    [3, 6, 9]]
                   
    """
    def __init__(self, dim=3, delay=1, skip=1):
        self._dim = dim
        self._delay = delay
        self._skip = skip

    def __call__(self, ts):
        """Transform method for single time-series data.
        Parameters
        ----------
        ts : list[float]
            A single time-series data.
        Returns
        -------
        point clouds : list[list[float, float, float]]
            Makes point cloud every a single time-series data.
        Raises
        -------
        TypeError
            If the parameter's type does not match the desired type.
        """
        ndts = np.array(ts)
        if ndts.ndim == 1:
            return self._transform(ndts)
        else:
            raise TypeError("Expects 1-dimensional array.")

    def fit(self, ts, y=None):
        return self
    
    def _transform(self, ts):
        """Guts of transform method."""
        return ts[
            np.add.outer(
                np.arange(0, len(ts)-self._delay*(self._dim-1), self._skip),
                np.arange(0, self._dim*self._delay, self._delay))
        ]

    def transform(self, ts):
        """Transform method for multiple time-series data.
        Parameters
        ----------
        ts : list[list[float]]
            Multiple time-series data.
        Attributes
        ----------
        ndts : 
             The ndts means that all time series need to have exactly 
             the same size. 
        Returns
        -------
        point clouds : list[list[list[float, float, float]]]
            Makes point cloud every a single time-series data.
        Raises
        -------
        TypeError
            If the parameter's type does not match the desired type.
        """
        ndts = np.array(ts)
        if ndts.ndim == 2:
            return np.apply_along_axis(self._transform, 1, ndts)
        else:
            raise TypeError("Expects 2-dimensional array.")
```